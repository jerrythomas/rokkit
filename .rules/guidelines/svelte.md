# Guidelines

- Use Svelte 5 runes mode for code
- Use SvelteKit for app
- Use UnoCSS for styling
- Use vitest for unit testing
- Use Playwright for e2e testing
- E2E tests should follow the page object pattern

## Svelte 5
{{[Svelte & Svelte Kit Distilled](https://svelte-llm.khromov.se/svelte-complete-distilled)}}
{{[svelte 5 documentation](https://svelte.dev/docs/svelte/llms.txt)}}

## Snippets

### Runes

Runes are special functions that replace reactivity declarations from Svelte 4.

```svelte
<script>
  // State
  let count = $state(0);

  // Derived values
  let doubled = $derived(count * 2);

  // Props
  let {
    name = 'world',
    value = $bindable(null)
  } = $props();

  // Effects
  $effect(() => {
    console.log(`Count changed to ${count}`);
  });
</script>
```

### Defining and passing snippets to components:

```svelte
<script>
  let items = ['one','two', 'three']
</script>
<!-- a header snippet -->
{#snippet header()}
  <h1>Custom header</h1>
{/snippet}

<!-- a snippet with parameters -->
{#snippet item(item, selected)}
  <div class:selected={selected}>
    {item.name}
  </div>
{/snippet}

<List {items} {header} {item} />
```

### Using snippets in a component:

```svelte
<script>
  let { header, item } = $props();
</script>

<div class="header">
  {@render header?.()}
</div>

{#each items as currentItem}
  {@render item(currentItem, selectedItem === currentItem)}
{/each}
```

To pass a snippet to another component:

#### ChildComponent.svelte

```svelte
<script>
  let { header } = $props();
</script>

{#if header}
  {@render header()}
{/if}
```

#### Snippets can be passed as properties to child component

```svelte
<script>
   import  ChildComponent from './ChildComponent.svelte'
</script>
{#snippet header()}
  <h1>Custom Header</h1>
{/snippet}

<ChildComponent {header} />
```

### Snippets can be enclosed within the child component.

```svelte
<script>
   import  ChildComponent from './ChildComponent.svelte'
</script>
<ChildComponent>
  {#snippet header()}
    <h1>Custom Header</h1>
  {/snippet}
</ChildComponent>
```
Both of these options work in a similar way

### Important: What NOT to do

Do not confuse snippets with React-style JSX or string templates:

```svelte
<!-- ❌ INCORRECT - Using arrow functions returning strings -->
<List
  header={() => `<h1>My List</h1>`}
  item={({ item }) => `<div>${item.name}</div>`}
/>

<!-- ✅ CORRECT - Using properly defined snippets -->
<!-- snippets are defined in the html section not in scripts block -->
{#snippet header()}
  <h1>My List</h1>
{/snippet}

{#snippet itemTemplate(item, selected)}
  <div class:selected={selected}>
    {item.name}
  </div>
{/snippet}

<List {header} item={itemTemplate} />
```

## Component Design Guidelines

- should be modular and data driven
- should be accessibile
- Reduced motion should be considered

## Project Structure

If the source code uses runes or the test uses runes the file extension should be `.svelte.js` or `.spec.svelte.js` respectively.

```
project/
│
├── src/
│   ├── index.js
│   ├── Component.svelte
│   └── utils.svelte.js
│
├── spec/
│   ├── index.spec.js
│   ├── Component.spec.js
│   └── utils.spec.svelte.js
├── e2e/
│   ├── home.e2e.js
│   └── login.e2e.js
└── README.md
```
