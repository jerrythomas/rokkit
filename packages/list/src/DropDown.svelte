<script>
	import List from './List.svelte'
	import ListItem from './ListItem.svelte'
	import Slider from './Slider.svelte'
	import DropDownIcon from './icons/DropDownIcon.svelte'

	export let data
	export let component = null
	export let selected = null
	export let key = 'id'
	export let selectedKey

	let opened = false

	function handleSelect(event) {
		console.log('selected', event)
		selected = event.detail
	}
</script>

<div
	class="flex flex-col outline-none w-full h-12 relative z-1 dropdown"
	class:opened
	on:focus={() => (opened = true)}
	on:blur={() => (opened = false)}
	tabindex={0}
>
	<div class="flex flex-shrink-0 h-12 items-center pl-4 selected-item">
		<span class="flex flex-grow">
			<ListItem {component} item={selected} />
		</span>
		<DropDownIcon />
	</div>

	{#if opened}
		<Slider class="h-60 absolute top-12 w-full z-1">
			<List
				bind:items={data}
				{key}
				{component}
				bind:selected={selectedKey}
				on:select={handleSelect}
				on:change
			/>
		</Slider>
	{/if}
</div>
