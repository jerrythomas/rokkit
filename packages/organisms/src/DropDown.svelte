<script>
	import { defaultFields, defaultStateIcons, createEmitter } from '@rokkit/core'
	import { dismissable } from '@rokkit/actions'
	import { Icon, Slider } from '@rokkit/atoms'
	import { Item } from '@rokkit/molecules'
	import List from './List.svelte'

	/**
	 * @typedef {Object} Props
	 * @property {string} [class]
	 * @property {any} [options]
	 * @property {import('@rokkit/core').FieldMapping} [fields]
	 * @property {any} [using]
	 * @property {any} [value]
	 * @property {any} [title]
	 * @property {any} [icon]
	 * @property {boolean} [small]
	 */

	/** @type {Props} */
	let {
		class: className = '',
		options = [],
		fields = $bindable(defaultFields),
		using = $bindable({ default: Item }),
		value = $bindable(null),
		title = null,
		icon = null,
		small = false,
		...events
	} = $props()

	// $effect.pre(() => {
	// using = { default: Item, ...using }
	// fields = { ...defaultFields, ...fields }
	// })
	let internalUsing = $derived({ default: Item, ...using })
	let internalFields = $derived({ ...defaultFields, ...fields })

	let emitter = createEmitter(events, ['change'])
	let offsetTop = $state(0)
	let open = $state(false)
	let icons = defaultStateIcons.selector

	function handleSelect(data) {
		open = false
		emitter.change(data)
	}
</script>

<!-- svelte-ignore a11y_no_noninteractive_tabindex -->
<drop-down
	class="dropdown relative flex cursor-pointer select-none {className}"
	class:open
	aria-haspopup="true"
	aria-controls="menu"
>
	<button
		class="flex items-center"
		bind:clientHeight={offsetTop}
		tabindex="0"
		use:dismissable
		ondismiss={() => (open = false)}
		onfocus={() => (open = true)}
		onblur={() => (open = false)}
	>
		<span class="flex items-center">
			{#if icon !== null}
				<Icon name={icon} />
			{/if}
			{#if !small && title}
				<p class="flex w-full">{title}</p>
			{/if}
			{#if open}
				<Icon name={icons.opened} />
			{:else}
				<Icon name={icons.closed} />
			{/if}
		</span>
	</button>
	{#if open}
		<Slider top={offsetTop + 4}>
			<List
				items={options}
				fields={internalFields}
				using={internalUsing}
				bind:value
				onselect={handleSelect}
				tabindex="-1"
			/>
		</Slider>
	{/if}
</drop-down>
