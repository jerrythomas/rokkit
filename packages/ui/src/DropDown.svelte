<script>
	import { defaultFields, defaultStateIcons } from '@rokkit/core'
	import { dismissable } from '@rokkit/actions'
	import Icon from './Icon.svelte'
	import Slider from './Slider.svelte'
	import Item from './Item.svelte'
	import List from './List.svelte'

	let {
		class: className,
		options = [],
		fields = defaultFields,
		using = { default: Item },
		value = null,
		title = null,
		icon = null,
		small = false,
		disabled = false
	} = $props()

	using = { default: Item, ...using }
	fields = { ...defaultFields, ...fields }

	let offsetTop = $state(0)
	let open = $state(false)
	let icons = defaultStateIcons.selector

	function handleSelect(event) {
		open = false
		// dispatch('change', event.detail)
	}
</script>

<drop-down
	class="dropdown relative flex cursor-pointer select-none {className}"
	class:open
	aria-haspopup="true"
	aria-controls="menu"
>
	<button
		class="flex items-center"
		bind:clientHeight={offsetTop}
		tabindex="0"
		use:dismissable
		ondismiss={() => (open = false)}
		onfocus={() => (open = true)}
		onblur={() => (open = false)}
	>
		<span class="flex items-center">
			{#if icon !== null}
				<Icon name={icon} />
			{/if}
			{#if !small && title}
				<p class="flex w-full">{title}</p>
			{/if}
			{#if open}
				<Icon name={icons.opened} />
			{:else}
				<Icon name={icons.closed} />
			{/if}
		</span>
	</button>
	{#if open}
		<Slider top={offsetTop + 4}>
			<List items={options} {fields} {using} bind:value onselect={handleSelect} tabindex="-1" />
		</Slider>
	{/if}
</drop-down>
