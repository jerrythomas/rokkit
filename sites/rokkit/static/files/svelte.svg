<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M18.2301 11.1217C21.1087 6.47773 15.3383 1.6935 11.2372 4.37297L7.337 6.85957C5.44834 8.08961 4.62832 10.453 5.78004 12.8817C2.90038 17.2501 8.50365 22.3655 12.7729 19.6332L16.6732 17.1466C18.7952 15.7494 19.2821 13.3398 18.2301 11.1217V11.1217ZM8.46536 8.46761L12.3599 5.98242C14.9184 4.45077 17.8644 7.12896 16.8023 9.62435C16.1715 9.12318 15.4685 8.81503 14.9345 8.67627C15.1375 7.87006 14.2016 7.27202 13.5507 7.68841L9.64763 10.1651C8.68258 10.8321 9.71148 12.204 10.6653 11.6284L12.1527 10.6803C15.4905 8.79676 18.6495 13.3192 15.5448 15.5357L11.6502 18.0209C9.30336 19.5215 6.10922 17.1175 7.20785 14.379C7.79145 14.8207 8.41131 15.1456 9.07564 15.3271C8.92896 16.2531 9.86333 16.7033 10.4694 16.3234L14.3626 13.8382C15.3233 13.1178 14.292 11.814 13.3464 12.375L11.8533 13.3231C8.55362 15.2514 5.32106 10.6843 8.46536 8.46761Z" fill="black"/>
</svg>
