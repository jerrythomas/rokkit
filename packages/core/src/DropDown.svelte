<script>
	import Icon from './Icon.svelte'
	import Text from './items/Text.svelte'
	import ListItems from './ListItems.svelte'
	import { defaultFields } from './constants.js'
	import { dismissable } from './actions'
	import { createEventDispatcher } from 'svelte'

	const dispatch = createEventDispatcher()

	let className = ''
	export { className as class }
	export let items = []
	export let fields = defaultFields
	export let using = { default: Text }
	export let value = null
	export let placeholder = 'Select a value'

	$: using = { default: Text, ...using }
	$: fields = { ...defaultFields, ...fields }

	let offsetTop = 0
	let open = false

	function handleSelect(event) {
		open = false
		dispatch('change', event.detail)
	}
</script>

<!-- svelte-ignore a11y-no-noninteractive-tabindex -->
<div
	class="flex w-full relative cursor-pointer dropdown {className}"
	class:open
	tabindex="0"
	use:dismissable
	on:blur={() => (open = false)}
	on:dismiss={() => (open = false)}
>
	<button
		on:click|stopPropagation={() => (open = !open)}
		class="flex w-full"
		bind:clientHeight={offsetTop}
		tabindex="-1"
	>
		<svelte:component this={using.default} content={value ?? placeholder} />
		<Icon name="dropdown-opened" />
	</button>
	{#if open}
		<div
			class="flex flex-col absolute z-10 h-fit w-full menu"
			style:top="{offsetTop}px"
		>
			<ListItems
				{items}
				{fields}
				{using}
				bind:activeItem={value}
				on:select={handleSelect}
			/>
		</div>
	{/if}
</div>
