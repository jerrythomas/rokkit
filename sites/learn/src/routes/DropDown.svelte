<script>
	import { equals } from 'ramda'
	import { FieldMapper } from '@rokkit/core'
	import { Item } from '@rokkit/ui'
	import { onMount } from 'svelte'

	let { value = $bindable(), options = [], mapping = new FieldMapper() } = $props()

	// let selectedOption = 'Source Type'
	let isOpen = $state(false)

	function toggleSelect() {
		isOpen = !isOpen
	}

	function selectOption(event) {
		value = options[event.currentTarget.dataset.index]
		// selectedOption = option.text
		isOpen = false
	}

	onMount(() => {
		document.addEventListener('click', handleClickOutside)
	})

	function handleClickOutside(event) {
		if (!event.target.closest('.custom-select')) {
			isOpen = false
		}
	}
</script>

<div class="custom-select-wrapper">
	<div class="custom-select" class:opened={isOpen}>
		<!-- svelte-ignore a11y_click_events_have_key_events -->
		<!-- svelte-ignore a11y_no_static_element_interactions -->
		<span class="custom-select-trigger" onclick={toggleSelect}>
			<item><Item {value} {mapping}></Item> </item>
		</span>
		<div class="custom-options">
			{#each options as option, index (index)}
				<!-- svelte-ignore a11y_click_events_have_key_events -->
				<!-- svelte-ignore a11y_no_static_element_interactions -->
				<rk-item aria-selected={equals(option, value)} data-index={index} onclick={selectOption}>
					<Item value={option} {mapping}></Item>
				</rk-item>
				<!-- <span class="custom-option {option.text === selectedOption ? 'selection' : ''}" on:click={() => selectOption(option)}>{option.text}</span> -->
			{/each}
		</div>
	</div>
</div>

<style>
	/** Custom Select **/
	.custom-select-wrapper {
		@apply relative inline-block min-w-40 select-none p-0;
		/* position: relative;
		display: inline-block;
		user-select: none; */
	}
	.custom-select {
		@apply relative inline-block w-full p-0;
		/* position: relative;
		display: inline-block; */
	}
	.custom-select-trigger {
		@apply relative block w-full cursor-pointer rounded-md;
		@apply flex w-full flex-row;
		@apply bg-blue-200 text-white;
		/* position: relative;
		display: block;
		width: 130px; */
		/* padding: 0 84px 0 22px;
		font-size: 22px;
		font-weight: 300;
		color: #fff;
		line-height: 60px;*/
		/* background: #5c9cd8; */
		/*border-radius: 4px;
		cursor: pointer; */
	}
	.custom-select-trigger:after {
		@apply -mt-3px content-'' border-b-1 border-r-1 absolute right-2 top-1/2 block h-10 w-10;
		@apply transform-rotate-45 translate-y-(-50%) transition-all-0.4s transform-origin-50%-0 ease-in-out;
		/* position: absolute;
		display: block;
		content: '';
		width: 10px;
		height: 10px;
		top: 50%;
		right: 25px; */
		/* margin-top: -3px; */
		/* border-bottom: 1px solid #fff;
		border-right: 1px solid #fff;
		transform: rotate(45deg) translateY(-50%);
		transition: all 0.4s ease-in-out;
		transform-origin: 50% 0; */
	}
	.custom-select.opened .custom-select-trigger:after {
		@apply transform-rotate-(-135deg) translate-y-(-50%);
		/* margin-top: 3px;
		transform: rotate(-135deg) translateY(-50%); */
	}
	.custom-options {
		@apply m-15px-0 border-1 shadow-0-2px-1px transition-all-0.4s-ease-in-out absolute left-0 right-0 top-full box-border block min-w-full rounded-md border-gray-300 bg-white;
		/* position: absolute;
		display: block;
		top: 100%;
		left: 0;
		right: 0;
		min-width: 100%;
		margin: 15px 0;
		border: 1px solid #b5b5b5;
		border-radius: 4px;
		box-sizing: border-box; */
		/* box-shadow: 0 2px 1px rgba(0, 0, 0, 0.07); */
		/* background: #fff; */
		/* transition: all 0.4s ease-in-out; */
		@apply visibility-hidden transform-translate-y-(-15px) pointer-events-none opacity-0;
		/* opacity: 0;
		visibility: hidden;
		pointer-events: none;
		transform: translateY(-15px); */
	}
	.custom-select.opened .custom-options {
		@apply visibility-visible pointer-events-all transform-translate-y-0;
		/* opacity: 1;
		visibility: visible;
		pointer-events: all;
		transform: translateY(0); */
	}
	.custom-options:before {
		@apply content-'' right-25px border-t-1 border-l-1;
		@apply transform-rotate-45 transition-all-0.4s-ease-in-out;
		@apply absolute bottom-full block bg-white;
		/* position: absolute;
		display: block;
		content: '';
		bottom: 100%;
		right: 25px;
		width: 7px;
		height: 7px;
		margin-bottom: -4px;
		border-top: 1px solid #b5b5b5;
		border-left: 1px solid #b5b5b5;
		background: #fff;
		transform: rotate(45deg);
		transition: all 0.4s ease-in-out; */
	}
	/* .option-hover:before {
		background: #f9f9f9;
	} */
	rk-item {
		@apply relative block min-h-10 w-full border-gray-300;
		@apply flex cursor-pointer flex-row items-center gap-2 text-lg text-gray-900;
		@apply transition-all-0.4s-ease-in-out;
		/* transition: all 0.4s ease-in-out; */
	}
	rk-item:first-of-type {
		@apply rounded-t-md;
	}
	rk-item:last-of-type {
		@apply border-bottom-none rounded-b-md;
		/* border-bottom: 0;
		border-radius: 0 0 4px 4px; */
	}
	rk-item:hover {
		@apply bg-gray-200;
		/* background: #f9f9f9; */
	}
	rk-item[aria-selected='true'] {
		@apply bg-red-300;
		/* background: #f9f9f9; */
	}
</style>
