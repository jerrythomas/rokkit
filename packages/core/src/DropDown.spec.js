import { describe, expect, it, beforeEach, vi, afterAll } from 'vitest'
import { cleanup, render, fireEvent } from '@testing-library/svelte'
import DropDown from '../src/DropDown.svelte'

describe('DropDown.svelte', () => {
	beforeEach(() => {
		cleanup()
	})
	afterAll(() => vi.resetAllMocks())

	it('should render the placeholder text by default', async () => {
		const { container } = render(DropDown, {
			props: { items: [], fields: {} }
		})
		const button = container.querySelector('.dropdown > button')
		expect(button.textContent.trim()).toBe('Select a value')
	})
	it('should toggle open state when button is clicked', async () => {
		const { container } = render(DropDown, {
			props: { items: [], fields: {} }
		})

		expect(container).toMatchSnapshot()
		const button = container.querySelector('.dropdown > button')
		await fireEvent.click(button)
		expect(button.parentNode.classList.contains('open')).toBe(true)
		await fireEvent.click(button)
		expect(button.parentNode.classList.contains('open')).toBe(false)
	})
	it('should close the dropdown when focus is lost', async () => {
		const { container } = render(DropDown, {
			props: { items: [], fields: {} }
		})
		const button = container.querySelector('.dropdown > button')
		await fireEvent.click(button)
		expect(button.parentNode.classList.contains('open')).toBe(true)
		await fireEvent.blur(button.parentNode)
		expect(button.parentNode.classList.contains('open')).toBe(false)
	})
	it('should dispatch a change event when an item is selected', async () => {
		let selected = null
		const items = [{ text: 'foo' }, { text: 'bar' }]

		const mockOnChange = vi.fn().mockImplementation((event) => {
			expect(event.detail).toEqual(selected)
		})
		const { container, component } = render(DropDown, {
			props: {
				items
			}
		})

		component.$on('change', mockOnChange)
		const button = container.querySelector('.dropdown > button')
		await fireEvent.click(button)
		expect(button.parentNode.classList.contains('open')).toBe(true)

		container
			.querySelectorAll('.dropdown list item')
			.forEach(async (item, index) => {
				selected = items[index]
				await fireEvent.click(item)
				expect(component.$$.ctx[component.$$.props.value]).toEqual(selected)
				expect(mockOnChange).toHaveBeenCalled()
			})
	})
})
