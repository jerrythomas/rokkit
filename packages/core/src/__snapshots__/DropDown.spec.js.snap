// Vitest Snapshot v1

exports[`DropDown.svelte > should toggle open state when button is clicked 1`] = `
<body>
  <div>
    <div
      class="flex w-full relative cursor-pointer select-none dropdown"
      tabindex="0"
    >
      <button
        class="flex w-full"
        tabindex="-1"
      >
         
         
        <p
          class="flex flex-grow"
        >
          Select a value
        </p>
         
        <square-icon
          class="flex flex-col flex-shrink-0 h-full items-center justify-center"
          style="width: 0px;"
        >
          <icon
            aria-hidden="true"
            class="selector-closed"
          />
          <iframe
            aria-hidden="true"
            src="about:blank"
            style="display: block; position: absolute; top: 0; left: 0; width: 100%; height: 100%; overflow: hidden; border: 0; opacity: 0; pointer-events: none; z-index: -1;"
            tabindex="-1"
          />
        </square-icon>
        <iframe
          aria-hidden="true"
          src="about:blank"
          style="display: block; position: absolute; top: 0; left: 0; width: 100%; height: 100%; overflow: hidden; border: 0; opacity: 0; pointer-events: none; z-index: -1;"
          tabindex="-1"
        />
      </button>
       
    </div>
  </div>
</body>
`;
